drivers/sensor/lis2dh/lis2dh.c

```diff
@@ -217,20 +217,25 @@ static int lis2dh_acc_odr_set(struct device *dev, u16_t freq)
 }
 #endif
 
-#ifdef CONFIG_LIS2DH_ACCEL_RANGE_RUNTIME
-static const union {
-	u32_t word_le32;
+/*
+ * Formula for the scale is:
+ *  ((SENSOR_G * 2 * (range_g)) / 65636LL)
+ */
+static const struct {
+	u16_t fs_scale[4];
 	u8_t fs_values[4];
-} lis2dh_acc_range_map = { .fs_values = {2, 4, 8, 16} };
+} lis2dh_acc_range_map = { .fs_scale = {583, 1168, 2335, 7007},
+			   .fs_values = {2, 4, 8, 16} };
 
+#ifdef CONFIG_LIS2DH_ACCEL_RANGE_RUNTIME
 static int lis2dh_range_to_reg_val(u16_t range)
 {
 	int i;
 	u32_t range_map;
 
-	range_map = sys_le32_to_cpu(lis2dh_acc_range_map.word_le32);
+	range_map = 16;
 
-	for (i = 0; range_map; i++, range_map >>= 1) {
+	for (i = 3; range_map; i--, range_map >>= 1) {
```

Comment:
Previous implementation was broken for sure, so this is an improvement. However I would further simplify it with something like:
```
#define LIS2DH_RANGE_IDX_TO_VALUE(idx)          (1 << ((idx) + 1))
#define LIS2DH_NUM_RANGES                       4

static int lis2dh_range_to_reg_val(u16_t range)
{
        int i;

        for (i = 0; i < LIS2DH_NUM_RANGES; i++) {
                if (range == LIS2DH_RANGE_IDX_TO_VALUE(i)) {
                        return i;
                }
        }

        return -EINVAL;
}
```